scio-core/src/main/java/com/spotify/scio/transforms/AsyncDoFn.java

```diff
@@ -0,0 +1,66 @@
+/*
+ * Copyright 2019 Spotify AB.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.spotify.scio.transforms;
+
+import com.google.common.cache.Cache;
+import com.google.common.collect.Maps;
+import java.util.UUID;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.Semaphore;
+import org.apache.beam.sdk.transforms.DoFn;
+import org.apache.beam.sdk.values.KV;
+
+public abstract class AsyncDoFn<A, B, C, D, FutureT> extends DoFn<A, KV<A, D>> {
```

Comment:
There's already a bunch of `AsyncDoFn`s and this will just add confusion. It should probably be named `BaseAsyncDoFn` to be consistent with the existing ones.